tests/numpy/test_np_reduction.py

```diff
@@ -85,10 +85,77 @@ def test_argops(nps_app_inst):
                 assert np.allclose(ba_result.get(), np_result)
 
 
+def test_average(nps_app_inst):
+    from nums import numpy as nps
+    from nums.numpy import BlockArray
+
+    assert nps_app_inst is not None
+
+    bas = [
+        nps.array([[[5, -2, 4, 8],
+                    [1, 2, 3, 4],
+                    [3, 2, 1, 0],
+                    [-1, -2, -3, 0]],
+                   [[6, -4, 2, 7],
+                    [4, 3, 2, 1],
+                    [1, 2, 0, 3],
+                    [0, -2, -3, -1]]]),
+        nps.array([[[5, -2, 4, 8],
+                    [1, 2, 3, 4],
+                    [3, 2, 1, 0],
+                    [-1, -2, -3, 0]],
+                   [[6, -4, 2, 7],
+                    [4, 3, 2, 1],
+                    [1, 2, 0, 3],
+                    [0, -2, -3, -1]]]),
+    ]
+    ba_wts = [
+        None,
+        nps.array([[[1, 2, 3, 4],
+                    [1, 2, 3, 4],
+                    [1, 2, 3, 4],
+                    [1, 2, 3, 4]],
+                   [[-2, -3, -4, -5],
+                    [-2, -3, -4, -5],
+                    [-2, -3, -4, -5],
+                    [-2, -3, -4, -5]]]),
+    ]
+    ba_shapes = [(1, 1, 1), (1, 2, 3), (2, 4, 3), (2, 4, 4)]
+    for ba, ba_wt in zip(bas, ba_wts):
+
+        for ba_shape in ba_shapes:
+            ba = ba.reshape(block_shape=ba_shape)
+            if ba_wt:
+                ba_wt = ba_wt.reshape(block_shape=ba_shape)
+            np_arr = ba.get()
+            np_wt = ba_wt.get() if ba_wt else None
+            op_params = ["average"]
+            axis_params = [None, 0, 1, 2]
+            returned = [True, False]
+
+            for op, axis, ret in itertools.product(op_params, axis_params, returned):
+                ns_op = nps.__getattribute__(op)
+                np_op = np.__getattribute__(op)
+                np_result = np_op(np_arr, axis=axis, weights=np_wt, returned=ret)
+                ba_result: BlockArray = ns_op(ba, axis=axis, weights=ba_wt, returned=ret)
+                if ret:
+                    np_avg, np_ws = np_result[0], np_result[1]
+                    ba_avg, ba_ws = ba_result[0], ba_result[1]
+                    assert np.allclose(ba_ws.get(), np_ws)
+                    assert ba_avg.grid.grid_shape == ba_avg.blocks.shape
+                    assert ba_avg.shape == np_avg.shape
+                    assert np.allclose(ba_avg.get(), np_avg)
+                else:
+                    assert ba_result.grid.grid_shape == ba_result.blocks.shape
+                    assert ba_result.shape == np_result.shape
+                    assert np.allclose(ba_result.get(), np_result)
+
```

Comment:
Just repeating here: Would be nice to add some tests that would trigger zeros in the weight sum. Doesn't need to be a part of the above loop.